### PR TITLE
GOES-16 and json data source transition related updates

### PR DESCRIPTION
**Pull request by malgosias**
Updates related to the transition to GOES-16 as the primary source of proton data and new (json) format of the data files. Affected files: ``get_hrc.py`` (major updates), ``plot_hrc.py`` (clean up).

### CODE REVIEW

File: get_hrc.py [hunk 1]
```diff
@@ -18,46 +18,65 @@
...
+    sec = DateTime(row['time_tag'][:-1], format='fits').secs
+    secs.append(sec)
+dat['time'] = secs
 
-secs = DateTime(mjd, format='mjd').secs
+descrs = [('time', 'f8'), ('hrc_shield', 'f8')]
+channels = ['P1', 'P2A', 'P2B', 'P3', 'P4', 'P5', 'P6',
+            'P7', 'P8A', 'P8B', 'P8C', 'P9', 'P10']
 
-hrc_shield = (6000 * dat['p4'] + 270000 * dat['p5']
```
Comment (jeanconn) [inline get_hrc.py:46]: Decided that we want to keep the old computation as a comment

Comment (jeanconn) [reply get_hrc.py:46]: I see the comments have been updated to describe this, but I think we really just want this bit of code in a comment.

**Comment by malgosias** [comment]
Ready for review. The new h5 file needed by the code can be found in ``/proj/sot/ska/data/arc3/hrc_shield_new.h5``. It ends on March 10, 2020, so we may consider running the new code for the first time with the 7-day json file, not the 6-hour json file as implemented.

File: get_hrc.py (hunk 1 above)
Comment (malgosias) [reply get_hrc.py:46]: Added pre-GOES-16 computations as comments.

File: get_hrc.py [hunk 2]
```diff
@@ -2,88 +2,158 @@
...
+def main():
+    parser = argparse.ArgumentParser(description='Archive GOES data and '
+                                     'HRC shield rate proxy')
+    parser.add_argument('--data-dir', type=str,
+                        default='.',
+                        help='Directory for output data files')
+    args = parser.parse_args()
+
+    dat = get_json_data()
+    newdat, descrs, hrc_bad = format_proton_data(dat)
```
Comment (jeanconn) [inline get_hrc.py:116]: It looks like format_proton_data uses the hrc_shield.h5 in the current working directory, which is not guaranteed to be a directory that contains hrc_shield.h5 .  There is a chdir later, but the chdir is too late and isn't a great idea anyway.

**Comment by jeanconn** [comment]
I tried to run an update to hrc_shield with the script and also ended up with values that are less than 0.1 . I think they are still legitimate even if we don't care (?) but they are flagged as bad by plot_hrc.py and the script errors out with nothing to plot.

File: get_hrc.py [hunk 3]
```diff
@@ -2,88 +2,158 @@
...
-        print(hrc_shield[ok].mean(), times[ok].mean(), file=f)
-
-for colname, scale, filename in zip(
-    ('p2', 'p5'), (3.3, 12.0), ('p4gm.dat', 'p41gm.dat')):
-    proxy = dat[colname][-3:] * scale
-    ok = proxy > 0
-    if len(proxy[ok]) > 0:
-        with open(filename, 'w') as f:
-            print(proxy[ok].mean(), times[ok].mean(), file=f)
+PATH = 'https://services.swpc.noaa.gov/json/goes/primary/'
```
Comment (taldcroft) [inline get_hrc.py:15]: More descriptive name like `URL_SWPC`.

**Comment by malgosias** [comment]
Introduced the style updates following our discussion and code and review comments for ``get_goes_x.py``:
- get_options()
- URL_NOAA, URL_7D, URL_6H
- Data gap detection
- **Most importantly:** conversion of fluxes from ‘per keV’ to ‘per MeV’. HRC shield values are at the level 2-3, as expected.

File: get_hrc.py [hunk 4]
```diff
@@ -1,89 +1,177 @@
...
+                                "HRC Antico shield + GOES", expectedrows=2e7)
+        h5.root.data.flush()
+
+    # Also write the mean of the last three values (15 minutes) to
+    # hrc_shield.dat.  Only include good values.
+    times = DateTime(newdat['time'][-3:]).unix
+    hrc_shield = newdat['hrc_shield'][-3:]
+    ok = ~hrc_bad[-3:]
+    if len(hrc_shield[ok]) > 0:
+        with open('hrc_shield.dat', 'w') as f:
```
Comment (jeanconn) [inline get_hrc.py:160]: Should this be in the 'data_dir'?  Otherwise I don't see data-dir being used.

File: get_hrc.py [hunk 5]
```diff
@@ -1,89 +1,177 @@
...
-        with open(filename, 'w') as f:
-            print(proxy[ok].mean(), times[ok].mean(), file=f)
+        with tables.open_file(args.h5, mode='r',
+                              filters=tables.Filters(complevel=5, complib='zlib')) as h5:
+            table = h5.root.data
+            descrs = table.dtype
+            lasttime = table.col('time')[-1]
+    except (OSError, IOError, tables.NoSuchNodeError):
+        print("Warning: No previous GOES shield data, using -1 as last time")
+        lasttime = -1
```
Comment (jeanconn) [inline get_hrc.py:131]: If we get in this part of the logic, `descrs` won't be defined and the `format_proton_data` below won't work.  Do we care?  Not sure if we should just exit here.

**Comment by jeanconn** [comment]
@malgosias It looks like I can't request review on your own PR, but I pushed a couple of commits to the end of this PR and it would be good for you to look them over and agree.  Thanks!

**Comment by jeanconn** [comment]
It looks like one issue that remains is that the current arc.pl page has logic to try to check the GOES proton table for non-zero values to see if things are up-to-date.  Both the code (wrong columns) and the logic (are zeros still valid?) need fixing.

https://github.com/sot/arc/blob/ad8d6fac53ff6ecdf9b4404c0947d536bf313491/arc.pl#L827

**Comment by jeanconn** [comment]
I think the question then is for @taldcroft and @malgosia.  Do we need a check for recent and valid proton data?  Should we update the arc.pl code to check for recent and valid data using the new hrc_shield.h5 file instead of using the proton channel html page?  Or should we update this previous check (which reads data from https://cxc.cfa.harvard.edu/mta/RADIATION_new/GOES/goes_pchan_p.html) to just read the columns used for the hrc proxy and check something reasonable in their values?   If so, what should the check be?

**Comment by taldcroft** [comment]
Presumably you can just fix the tests?  It is fairly important to be aware of stale data on the page since this does happen and it is easy to overlook.

If I recall my perl, the `@{$p2} == 0` is checking for a zero-length list. So if there is some easy translation to the new column names then it should be OK.  And I've just forgotten the translation between P2, P5 and P4GM and P41GM.

I note also that in the replan central page, P41GM is still black (not greyed out) wrongly implying that it is an on-board active monitor. (Though maybe MTA is yellow-alerting on that??)  That should be fixed, and maybe the labels should be made more clear by adding "proxy" to both of them as a reminder that these are GOES proxies for old EPHIN channels.

**Comment by taldcroft** [comment]
I've heard from a back-channel that @malgosias is going to be otherwise busy soon for a few weeks, so we should try to wrap this up soon!

**Comment by jeanconn** [comment]
Should we just add a date to hrc_shield.dat and check that?

**Comment by jeanconn** [comment]
I could probably also use a new issue for P41GM.  Am I supposed to take an action to figure out if mta is alerting on it before greying it out too?  Or can we just remove these from the table at this point (were we loathe to before because somebody or something was parsing the table?).

File: get_hrc.py [hunk 6]
```diff
@@ -1,89 +1,179 @@
...
+    ok = ~hrc_bad[-3:]
+    if len(hrc_shield[ok]) > 0:
+        Path(args.data_dir).mkdir(exist_ok=True)
+        with open(Path(args.data_dir, 'hrc_shield.dat'), 'w') as f:
+            print(hrc_shield[ok].mean(), times[ok].mean(), file=f)
+
+    # For GOES earlier than 16:
+    # for colname, scale, filename in zip(
+    #     ('p2', 'p5'), (3.3, 12.0), ('p4gm.dat', 'p41gm.dat')):
+    # GOES-16, ``scale`` TBD
```
Comment (jeanconn) [inline get_hrc.py:168]: I suppose I don't really understand this TBD comment

**Comment by jeanconn** [comment]
It looks like the second value in hrc_shield.gat *is* a mean time of the last three "valid" samples.   Seems like a staleness check on that would be the right thing then, as we're not using the proton html page for the hrc shield values.  

When do we want a warning it is stale?  How old should it be?

File: arc.pl [hunk 7]
```diff
@@ -809,32 +808,33 @@ sub make_ephin_goes_table {
     my %val;
     my %tab_def = %{$opt{ephin_goes_table}};
 
-    my $start = qr/P1 \s+ P2  \s+ P5 \s+ P8  \s+ P10 \s+ P11 \s+ H2/x;
-
-    my ($goes_date, $p2, $p5);
```
Comment (jeanconn) [inline arc.pl:814]: I figured there was really no reason to *parse* the MTA proton page at this point.

File: arc.pl [hunk 8]
```diff
@@ -809,32 +808,33 @@ sub make_ephin_goes_table {
...
     my $p4gm_delta = Event::calc_delta_date($p4gm_time);
     my $p41gm_delta = Event::calc_delta_date($p41gm_time);
 
-    $val{GOES}{P4GM}  = (defined $p2 and @{$p2}) ? sprintf("%.2f", $p4gm_proxy) : '---'; # See http://asc.harvard.edu/mta/G10.html
-    $val{GOES}{P41GM} = (defined $p5 and @{$p5}) ? sprintf("%.2f", $p41gm_proxy) : '---'; # ditto
+    # Add a warning above the CXO GOES rates table if the proton data for these
+    # is stale.  Since these times are all from the proton/get_hrc data they
+    # should be the same, but that implementation could change.
+    my $warning = '<h2 style="color:red;text-align:center">';
+    if (($CurrentTime - $hrc_time) > 86400){
```
Comment (jeanconn) [inline arc.pl:825]: We were already printing the "delta" time for each of these metrics underneath the table, so these are just red warnings above the table to call attention to stale data.  I figured that would be more obvious than updating the font color in the fine print under/within the CXO GOES rates table.  I tested this by just setting the 86400 to a really small value and confirming expected deltas and warning.

File: get_hrc.py (hunk 6 above)
Comment (jeanconn) [reply get_hrc.py:168]: I still don't know what the TDB should be but guessing we don't care.

File: arc.pl (hunk 8 above)
Comment (jeanconn) [reply arc.pl:825]: @taldcroft if you agree this approach is somewhat acceptable, I think we should be done with this PR for now (we need future feedback on P41GM and other limits but probably doesn't need to be in here).

**Pull request merged**